docs/topics/js/js-ir-migration.md

```diff
@@ -0,0 +1,186 @@
+in expressions without being defined. 
+
+```kotlin
+external interface ComponentProps: RProps {
+    var isInitialized: Boolean
+    var visible: Boolean
+}
+```
+
+```kotlin
+val props = js("{}") as ComponentProps
+props.isInitialized = true
+// visible is not initialized - OK in JS – means it's false
+if (props.isInitialized && props.visible) {} // not OK in Kotlin
+```
+
+However, this code with throw a `ClassCastException` because `visible` is `null` while it's type doesn't allow this.
+
+**Solution**: make all `Boolean` properties of external interfaces nullable (`Boolean?`):
+
+```kotlin
+// Replace this
+external interface ComponentProps: RProps {
+    var visible: Boolean
+}
+```
+
+```kotlin
+// With this
+external interface ComponentProps: RProps {
+    var visible: Boolean?
+}
+```
+
+## Convert functions with receiver in external interfaces to regular functions 
+
+**Issue**: external declarations can't contain functions with receiver, such as extension functions or properties with corresponding
+functional types.
+
+**Solution**: convert such functions and properties to regular functions by adding the receiver object as an argument:
+
+```kotlin
+// Replace this
+external interface ButtonProps : RProps {
+    var inside: StyledDOMBuilder<BUTTON>.() -> Unit
+}
+```
+
+```kotlin
+external interface ButtonProps : RProps {
+    var inside: (StyledDOMBuilder<BUTTON>) -> Unit
+}
+```
+
+## Create plain JS objects for interoperability
+
+**Issue**: Properties of a Kotlin object that implements an external interface are not _enumerable_. This means that they are not 
+visible for operations that iterate over the object's properties, such as:
+* `for (var name in obj)`
+* `console.log(obj)`
+* `JSON.stringify(obj)`
+* And other
+
+Although they are still accessible by the name: `obj.myProperty`
+
+```kotlin
+external interface AppProps { var name: String }
+data class AppPropsImpl(override var name: String) : AppProps
+fun main() {
+    val jsApp = js("{name: 'App1'}") as AppProps // plain JS object
+    println("Kotlin sees: ${jsApp.name}") // "App1"
+    println("JSON.stringify sees:" + JSON.stringify(jsApp)) // {"name":"App1"} - OK
+
+    val ktApp = AppPropsImpl("App2") // Kotlin object
+    println("Kotlin sees: ${ktApp.name}") // "App2"
+    // JSON sees only the backing field, not the property
+    println("JSON.stringify sees:" + JSON.stringify(ktApp)) // {"_name_3":"App2"} 
+}
+```
+
+**Solution 1**: create plain JavaScript objects with `js()` or `jsObject()`:
+
+```kotlin
+external interface AppProps { var name: String }
+data class AppPropsImpl(override var name: String) : AppProps
+```
+
+```kotlin
+// Replace this
+val ktApp = AppPropsImpl("App2") // Kotlin object
+```
+
+```kotlin
+// With this
+val jsApp = js("{name: 'App1'}") as AppProps // or jsObject {}
+```
+
+**Solution 2**: create objects with `kotlin.js.json()`:
+
+```kotlin
+val jsonApp = kotlin.js.json(Pair("name", "App3")) as AppProps
```

Comment:
Is this a good idea? I'm not judging, I just haven't ever seen this before

Reply:
@ilgonmic what do you think?

Reply:
I saw this earlier in user projects, and it will work
But don’t know which way will be preferable for users to create plain JS objects
This one is good, because not require additional dependencies